### PR TITLE
Provide internals test utilities

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Add internals accessor helpers

Introduces ITestAccessor and related classes to keep test specific code out of the shipping assemblies.

Allows easy dynamic access to private/internal fields and members and provides a pattern for writing typed accessors.

Also supports creating of delegates, which allows for access to members that utilize ref structs which are not yet supported in reflection.

Removes existing TestAccessor classes from the shipping code. Updates most of the internal virtual methods on TextBox to private protected to encapsulate. Various comment / line spacing fixes to TextBox.

There are three commits in this PR. The first two clean up tests and internal access for the EventHandlerService. I kept them separate as they don't depend on the new functionality as I removed all of the internals access there.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3225)

### CODE REVIEW

**Comment by RussKie** [comment]
Tests are failing with the likes of:
```
    System.Windows.Forms.Tests.TextBoxTests.TextBox_ShouldRenderPlaceHolderText(textBox: System.Windows.Forms.Tests.TextBoxTests+SubTextBox, Text: , m: msg=0xf (WM_PAINT) hwnd=0x0 wparam=0x0 lparam=0x0 result=0x0, expected: False) [FAIL]
      Microsoft.CSharp.RuntimeBinder.RuntimeBinderException : 'System.Dynamic.DynamicObject' does not contain a definition for 'ShouldRenderPlaceHolderText'
      Stack Trace:
           at CallSite.Target(Closure , CallSite , Object , Message )
           at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
        /_/src/System.Windows.Forms/tests/UnitTests/TextBoxTests.cs(380,0): at System.Windows.Forms.Tests.TextBoxTests.TextBox_ShouldRenderPlaceHolderText(TextBox textBox, Message m, Boolean expected)
```

Please also rebase on top of the latest master.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 1]
```diff
@@ -533,7 +518,7 @@ protected override void Dispose(bool disposing)
                 // Reset this just in case, because the SHAutoComplete stuff
                 // will subclass this guys wndproc (and nativewindow can't know about it).
                 // so this will undo it, but on a dispose we'll be Destroying the window anyay.
-                //
+
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs:521]: [nit] not needed
```suggestion
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs [hunk 2]
```diff
@@ -805,8 +805,7 @@ private void WmVScroll(ref Message m)
         /// <summary>
         ///  Handles the WM_KEYDOWN message.
         /// </summary>
-        //added to handle keyboard events
-        //
+
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewControl.cs:808]: ditto
```suggestion
```

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 3]
```diff
@@ -299,9 +289,12 @@ protected override CreateParams CreateParams
                 }
 
                 // Translate for Rtl if necessary
-                //
+
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs:292]: ditto
```suggestion
```

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBoxBase.cs [hunk 4]
```diff
@@ -845,13 +845,7 @@ public new Padding Padding
...
         ///  have to care about this.
         /// </summary>
-        virtual internal bool PasswordProtect
-        {
-            get
-            {
-                return false;
-            }
-        }
+        virtual private protected bool PasswordProtect => false;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBoxBase.cs:848]:
```suggestion
        private protected virtual bool PasswordProtect => false;
```

File: src/System.Windows.Forms/tests/UnitTests/TestAccessorTests.cs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            Assert.Equal(1970, access.Long);
+        }
+
+        [Fact]
+        public void TestAccessor_DynamicAccess_StaticProperty()
+        {
+            var testClass = new PrivateTestClass();
+            dynamic access = testClass.TestAccessor().Dynamic;
+            access.Int = 1989;
+            Assert.Equal(1989, access.Int);
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/TestAccessorTests.cs:78]: This is spooky - accessing a static via an instance.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/EventHandlerService.cs [hunk 6]
```diff
@@ -84,19 +83,18 @@ public void PopHandler(object handler)
                 throw new ArgumentNullException(nameof(handler));
             }
 
-            for (HandlerEntry entry = _handlerHead; entry != null; entry = entry.next)
+            var node = _handlers.Find(handler);
+            if (node != null)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/EventHandlerService.cs:87]: I would invert if and return

File: src/System.Windows.Forms.Design/tests/UnitTests/System/Windows/Forms/Design/EventHandlerServiceTests.cs [hunk 7]
```diff
@@ -40,72 +28,66 @@ public void GetHandler_should_throw_if_handlerType_null()
...
-        {
-            var service = new EventHandlerService(null);
-            var handler = new Button();
-            service.GetTestAccessor().LastHandlerType = typeof(Button);
-            service.GetTestAccessor().LastHandler = handler;
-
-            Assert.Same(handler, service.GetHandler(typeof(Button)));
-        }
+        private class A { }
+        private class B : A { }
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/System/Windows/Forms/Design/EventHandlerServiceTests.cs:39]: My convention we keep test mocks and stubs at the end of the file, makes it easier to navigate

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs:521]: I prefer to keep a line break between blocks of text and code to make it easier to read (doesn't runalltogether).

File: src/System.Windows.Forms/tests/UnitTests/TestAccessorTests.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/tests/UnitTests/TestAccessorTests.cs:78]: A little odd, but it isn't using the `this` pointer outside of finding the `Type` with it. It didn't make sense to me to make people jump through hoops of creating two accessors when they start with an instance.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBoxBase.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBoxBase.cs:848]: There has got to be a rule for this somewhere. Would be nice to turn it on if there is.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/EventHandlerService.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/EventHandlerService.cs:87]: I typically do as well, I hesitated at first as I was trying to keep the diff simpler, but it doesn't change much here.

Comment (JeremyKuhne) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/EventHandlerService.cs:87]: Hmm- inverting here is a little less obvious due to the assert. I really don't like the assert here- I'm just going to kill it.

**Comment by RussKie** [comment]
Few issues:
1. Need to rebase on top of the latest
2. The wrapper is not retrieving the correct type - it returns `System.Windows.Forms.Tests.TextBoxTests.SubTextBox` instead of `System.Windows.Forms.TextBox` and this causes tests to fail
    image

**Pull request merged**